### PR TITLE
Add target_oid and upstream attributes to BranchInfo

### LINKED ISSUE #962: refactor: add target_oid and upstream attributes to BranchInfo
## Summary

Enhance `BranchInfo` to include two additional attributes:
- `target_oid` - the commit object ID the branch points to
- `upstream` - the configured upstream/tracking branch (as a `BranchInfo` object)

## Current State

```ruby
BranchInfo = Data.define(:refname, :current, :worktree, :symref)
```

## Proposed State

```ruby
BranchInfo = Data.define(:refname, :target_oid, :current, :worktree, :symref, :upstream)
```

### Attribute: `target_oid`

- **Name**: Use `target_oid` (not `sha` or `objectname`)
  - `oid` is git's official term for "object ID" and is algorithm-agnostic (future-proof for SHA-256 transition)
  - `target_` prefix clarifies this is what the branch points to (branches aren't objects themselves)
  - Aligns with libgit2/rugged conventions
  - Consistent with planned `TagInfo` update (see #963)
- **Type**: `String` - the full 40-character object ID
- **Source**: `%(objectname)` format placeholder in `git branch --format`

### Attribute: `upstream`

- **Type**: `BranchInfo` or `nil`
- **Purpose**: Represents the configured upstream/tracking branch
- **Source**: `%(upstream)` and `%(upstream:objectname)` format placeholders

The upstream is modeled as a `BranchInfo` object because:
1. The upstream is usually a remote-tracking branch (e.g., `origin/main`), but can also be a local branch
2. Using `BranchInfo` provides type consistency and allows `branch.upstream.remote_name`, `branch.upstream.short_name` to work naturally
3. Remote-tracking branches have `upstream: nil`, providing natural recursion termination

### Upstream examples

**Local branch tracking a remote-tracking branch (most common):**
```ruby
BranchInfo.new(
  refname: 'main',
  target_oid: 'abc123...',
  current: true,
  worktree: false,
  symref: nil,
  upstream: BranchInfo.new(
    refname: 'remotes/origin/main',
    target_oid: 'abc123...',
    current: false,
    worktree: false,
    symref: nil,
    upstream: nil  # remote-tracking branches don't have upstreams
  )
)
```

**Local branch tracking another local branch (rare but valid):**
```ruby
BranchInfo.new(
  refname: 'feature',
  target_oid: 'def456...',
  current: false,
  worktree: false,
  symref: nil,
  upstream: BranchInfo.new(
    refname: 'main',
    target_oid: 'abc123...',
    current: false,
    worktree: false,
    symref: nil,
    upstream: nil  # or could have its own upstream
  )
)
```

**Branch with no upstream configured:**
```ruby
BranchInfo.new(
  refname: 'orphan-branch',
  target_oid: 'ghi789...',
  current: false,
  worktree: false,
  symref: nil,
  upstream: nil
)
```

## Git command for retrieving data

All data can be retrieved in a single `git branch --format` call:

```bash
git branch --format='%(refname) %(objectname) %(HEAD) %(worktreepath) %(symref) %(upstream) %(upstream:objectname)'
```

## Files to Modify

1. **`lib/git/branch_info.rb`**
   - Add `target_oid` and `upstream` to `Data.define`
   - Update documentation and examples

2. **`lib/git/commands/branch/list.rb`**
   - Update `--format` string to include `%(objectname)`, `%(upstream)`, `%(upstream:objectname)`
   - Update parsing logic to construct upstream `BranchInfo` objects
   - Handle cases where upstream doesn't exist or upstream OID is unavailable

3. **`lib/git/commands/branch/create.rb`**
   - Update to provide `target_oid` and `upstream` when constructing `BranchInfo`
   - Note: newly created branches typically don't have an upstream unless `--track` is used

4. **All call sites constructing `BranchInfo`** - search for `BranchInfo.new(` and update:
   - `lib/git/remote.rb`
   - `lib/git/base.rb`
   - Any other locations

5. **Tests**
   - Update existing `BranchInfo` tests
   - Add tests for upstream scenarios (remote-tracking, local, nil)
   - Update `Branch::List` tests for new format parsing
   - Update `Branch::Create` tests

## Related Issues

- #963 - Adding `oid` and `target_oid` to `TagInfo` (uses same naming convention)

## Notes

- For remote-tracking branches listed via `git branch -r`, the upstream will always be `nil`
- When upstream exists but the remote-tracking branch hasn't been fetched, `%(upstream:objectname)` may be empty—handle this gracefully (set `target_oid: nil` on the upstream `BranchInfo`)
- Don't consider this a breaking change in the commit or the PR since `BranchInfo` hasn't been released yet

### PR DESCRIPTION
**Pull request by jcouball**
## Summary

Implements #962 by adding `target_oid` and `upstream` attributes to `BranchInfo`.

### BranchInfo Data Structure
- Added `target_oid` attribute - the full SHA of the commit the branch points to
- Added `upstream` attribute - a `BranchInfo` object for the upstream tracking branch
- Added YARD documentation for all attributes

### Branch::List Command
- Added `FORMAT_STRING` constant using `git branch --format` placeholders
- Format: `%(refname:short)|%(objectname)|%(HEAD)|%(worktreepath)|%(symref)|%(upstream)`
- Replaced regex-based parsing with format-based parsing
- `target_oid` is now populated from `%(objectname)`
- `upstream` is now populated as a `BranchInfo` from `%(upstream)`

### Call Sites Updated
- `lib/git/base.rb` - Updated `BranchInfo.new` call
- `lib/git/remote.rb` - Updated `BranchInfo.new` call
- `lib/git/commands/branch/create.rb` - Returns nil for `target_oid` and `upstream` (newly created branches)

### Tests
- Unit tests for BranchInfo with new attributes
- Unit tests for Branch::List format parsing
- Integration test for upstream tracking with a real remote

## Note on `%(upstream:objectname)`

The issue originally suggested using `%(upstream:objectname)` to get the upstream's commit SHA. However, this placeholder is not supported in git. The `upstream.target_oid` will be `nil` - users who need the upstream's OID can look up the upstream branch separately.

Closes #962

### CODE REVIEW

File: lib/git/commands/branch/list.rb [hunk 1]
```diff
@@ -131,94 +118,119 @@ def initialize(execution_context)
...
-        def format_branch_data(match_data)
-          Git::BranchInfo.new(
-            refname: match_data[:refname],
-            current: !match_data[:current].nil?,
-            worktree: !match_data[:worktree].nil?,
-            symref: match_data[:symref]
-          )
+        def detached_head?(refname)
+          refname.start_with?('(')
+        end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/commands/branch/list.rb]: Renamed `detached_head?` to `non_branch_entry?` - Now uses specific regex patterns ^\\(HEAD detached and ^\\(not a branch\\) instead of matching any refname starting with (. This ensures legitimate branch names starting with ( won't be filtered out.

File: lib/git/commands/branch/list.rb [hunk 2]
```diff
@@ -131,94 +118,119 @@ def initialize(execution_context)
...
-            Line #{index + 1}:
-              "#{line}"
-          ERROR
+          Git::BranchInfo.new(
+            refname: refname,
+            target_oid: nil, # We don't have upstream's OID from this format
+            current: false,
+            worktree: false,
+            symref: nil,
+            upstream: nil # Upstream branches don't have their own upstream in this context
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/commands/branch/list.rb:224]: `build_upstream_info` now reuses `normalize_refname` - Removed the duplicate `sub(%r{^refs/}, '')` logic and now calls the shared `normalize_refname` helper for consistent handling of both local (refs/heads/main → main) and remote (refs/remotes/origin/main → remotes/origin/main) upstream refs.

File: lib/git/commands/branch/list.rb [hunk 3]
```diff
@@ -131,94 +118,120 @@ def initialize(execution_context)
...
-            current: !match_data[:current].nil?,
-            worktree: !match_data[:worktree].nil?,
-            symref: match_data[:symref]
-          )
+        # @param refname [String] the refname to check
+        # @return [Boolean] true if this is a non-branch entry
+        #
+        def non_branch_entry?(refname)
+          refname.match?(/^\(HEAD detached/) || refname.match?(/^\(not a branch\)/)
+        end
```
[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/commands/branch/list.rb:181]: Make sure there are rspec integration tests for both detached head and not a branch cases.

[the remark this answers is not in the document]
Comment (jcouball) [reply lib/git/commands/branch/list.rb:181]: Added the following:

1. Unit test for "(not a branch)" filtering - in `list_spec.rb`
2. Integration test for detached HEAD - in `list_spec.rb` - creates a real repo, checks out a tag to enter detached HEAD state, and verifies only real branches are returned
